gfw_pixetl/layer.py

```diff
@@ -523,45 +527,93 @@ def transform(self, tiles: Iterator[RasterSrcTile]) -> Iterator[RasterSrcTile]:
+            # o, e = p.communicate()
+            #
+            # if p.returncode != 0:
+            #     logger.error("Could not calculate file " + tile.calc_uri)
+            #     logger.exception(e)
+            #     raise GDALError(e)
+            try:
+                self._calc(tile)
+            except Exception:
+                logger.exception("Calculation failed")
+                raise
             else:
-                no_data_cmd = list()
+                yield tile
 
-            cmd: List[str] = (
-                ["gdal_calc.py", "--type", self.data_type.data_type]
-                + no_data_cmd
-                + [
-                    "-A",
-                    tile.calc_uri,
-                    "--calc={}".format(self.calc),
-                    "--outfile={}".format(tile.uri),
-                    "--co",
-                    "COMPRESS={}".format(self.data_type.compression),
-                    "--co",
-                    "NBITS={}".format(self.data_type.nbits),
-                    "--co",
-                    "TILED=YES",
-                    "--co",
-                    "BLOCKXSIZE={}".format(tile.grid.blockxsize),
-                    "--co",
-                    "BLOCKYSIZE={}".format(tile.grid.blockysize),
-                    "--quiet",
-                ]
-            )
+    def _calc(self, tile):
+        with rasterio.Env(GDAL_TIFF_INTERNAL_MASK=True):
+            with rasterio.open(tile.calc_uri) as src:
+
+                kwargs = src.meta.copy()
+                kwargs.update(
+                    {
+                        "dtype": self.data_type.to_numpy_dt(),
+                        "compress": self.data_type.compression,
+                        "tiled": True,
+                        "blockxsize": self.grid.blockxsize,
+                        "blockysize": self.grid.blockysize,
+                    }
+                )
+                if self.data_type.no_data == 0 or self.data_type.no_data:
+                    kwargs.update({"nodata": self.data_type.no_data})
+                else:
+                    kwargs.update({"nodata": None})
 
-            logger.info("Calculate tile " + tile.tile_id)
-            p = sp.Popen(cmd, stdout=sp.PIPE, stderr=sp.PIPE)
-            o, e = p.communicate()
+                if self.data_type.nbits:
+                    kwargs.update({"nbits": self.data_type.nbits})
 
-            if p.returncode != 0:
-                logger.error("Could not calculate file " + tile.calc_uri)
-                logger.exception(e)
-                raise GDALError(e)
-            else:
-                yield tile
+                with rasterio.open(tile.uri, "w", **kwargs) as dst:
+
+                    for block_index, window in src.block_windows(1):
+                        data = src.read(window=window, masked=True)
+
+                        # apply user submitted calculation
+                        funcstr = "def f(A):\n    return {e}".format(e=self.calc)
```

Comment:
Instead of gdalcalc use rasterio/numpy to change pixel values.
Read/ write data block by block to keep memory footprint low.
We can pass the numpy function as string and convert to Callable using exec()